File: Behavioral design pattern/command_design_pattern.cpp
#include "../bitsStd.h"
using namespace std;


//receiver
class airConditioner {
bool isOn = false;
public:
void turnOn() {
    isOn = true;
    cout << "AC is turned ON!" << endl;
}

void turnOff() {
    isOn = false;
    cout << "AC is turned OFF!" << endl;
}
};

//command interface
class ICommand {
public:
virtual void execute() = 0;
virtual void undo() = 0;
};

class TurnOffAC : public ICommand {
airConditioner* ac;
public:
TurnOffAC(airConditioner* ac) {
    this->ac = ac;
}

void execute() {
ac->turnOff();
}

void undo() {
ac->turnOn();
}
};

class TurnOnAC : public ICommand {
airConditioner* ac;
public:
TurnOnAC(airConditioner* ac) {
    this->ac = ac;
}

void execute() {
ac->turnOn();
}

void undo() {
ac->turnOff();
}
};

//invoker
class Remote {
stack<ICommand*>remote_stack;
ICommand* command;
public:

void setCommand(ICommand* command) {
    this->command = command;
    remote_stack.push(command);
}

void pressButton() {
    command->execute();
}

void undo() {
    if(!remote_stack.empty()) {
    ICommand* topCommand = remote_stack.top();
    remote_stack.pop();
    topCommand->undo();
    } else {
        cout << "No command that can be undone!" << endl;
    }
}
};

int main() {
airConditioner* ac = new airConditioner();
Remote* remote = new Remote();

remote->setCommand(new TurnOnAC(ac));
remote->pressButton();

remote->setCommand(new TurnOffAC(ac));
remote->pressButton();

remote->undo(); // ac turn off should be undone
remote->undo(); //ac turn on should be undone
remote->undo(); // should throw exception
}

File: Design_qsns/Design_patterns_java/Behavioral_design_pattern/command_design_pattern.java
//"Command pattern lets you wrap a request (or action) as an object, so you can store it, pass it around, delay it, or undo it later."
package Design_qsns.Design_patterns_java.Behavioral_design_pattern;
import java.util.*;

class AirConditioner { // receiver
    public void turnOn() {
        System.out.println("Turn on the AC");
    }

    public void turnOff() {
        System.out.println("Turn off the AC");
    }
}

interface ICommand { // command interface
    void execute();
    void undo();
}

class TurnOnCommand implements ICommand { // concrete commands
    AirConditioner ac;
    TurnOnCommand(AirConditioner ac) {
        this.ac = ac;
    }
    public void execute() {
        ac.turnOn();
    }

    public void undo() {
      ac.turnOff();  
    }
}

class TurnOffCommand implements ICommand {
    AirConditioner ac;
    TurnOffCommand(AirConditioner ac) {
        this.ac = ac;
    }
    public void execute() {
        ac.turnOff();
    }

    public void undo() {
      ac.turnOn();  
    }
}

class Remote { // invoker
    Stack<ICommand> commandHistory;
    ICommand command;
    Remote() {
        this.commandHistory = new Stack<>();
    }

    public void setCommand(ICommand command) {
        this.command = command;
        commandHistory.push(command);
    }

    public void pressButton() {
        command.execute();
    } 
    
    public void undo() {
        if(!commandHistory.isEmpty()) {
            ICommand topCommand = commandHistory.peek();
            topCommand.undo();
            commandHistory.pop();
        } else {
            System.out.println("No commands to undo");
        }
    }
}

public class command_design_pattern {
    public static void main(String[] args) {
       AirConditioner ac = new AirConditioner(); // receiver is created
       Remote remote = new Remote(); //remote that is invoker is created
       ICommand command_on = new TurnOnCommand(ac); // command created for turn on that takes ac on which to pass the command
       ICommand command_off = new TurnOffCommand(ac);
       remote.setCommand(command_on); // now button decided for remote , and once button decided command is set
       remote.pressButton();    //once button pressed , set command is executed
       remote.setCommand(command_off);
       remote.pressButton();
       remote.undo();   // this helps to undo just the last command
       remote.undo();
    }
}
